controller/kongplugininstallation/controller.go

```diff
@@ -0,0 +1,219 @@
+		}
+		if err := ctrl.SetControllerReference(&kpi, &cm, r.Scheme); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.Client.Create(ctx, &cm); err != nil {
+			return ctrl.Result{}, err
+		}
+		kpi.Status.UnderlyingConfigMapName = cm.Name
+	case 1:
+		cm = cms[0]
+		cm.Data = map[string]string{
+			fmt.Sprintf("%s.lua", kpi.Name): string(plugin),
+		}
+		if err := r.Client.Update(ctx, &cm); err != nil {
+			return ctrl.Result{}, err
+		}
+	default:
+		// It should never happen.
+		return ctrl.Result{}, errors.New("unexpected error happened - more than one ConfigMap found")
+	}
+
+	return ctrl.Result{}, setStatusConditionForKongPluginInstallation(
+		ctx, r.Client, &kpi, metav1.ConditionTrue, v1alpha1.KongPluginInstallationReasonReady, "plugin successfully saved in cluster as ConfigMap",
+	)
+}
+
+func (r *Reconciler) listKongPluginInstallationsForSecret(ctx context.Context, obj client.Object) []reconcile.Request {
+	name, namespace := obj.GetName(), obj.GetNamespace()
+
+	var kpiList v1alpha1.KongPluginInstallationList
+	if err := r.List(ctx, &kpiList); err != nil {
+		k8slog.FromContext(ctx).Error(
+			err,
+			"failed to run map funcs for secrets",
+		)
+		return nil
+	}
+
+	var recs []reconcile.Request
+	for _, kpi := range kpiList.Items {
+		if kpi.Spec.ImagePullSecretRef == nil {
+			continue
+		}
+		if kpi.Spec.ImagePullSecretRef.Namespace == "" {
+			kpi.Spec.ImagePullSecretRef.Namespace = kpi.Namespace
+		}
+		if kpi.Spec.ImagePullSecretRef.Namespace == namespace && kpi.Spec.ImagePullSecretRef.Name == name {
+			recs = append(recs, reconcile.Request{
+				NamespacedName: client.ObjectKey{
+					Name:      kpi.Name,
+					Namespace: kpi.Namespace,
+				},
+			})
+		}
+	}
+	return recs
+}
+
+func setStatusConditionFailedForKongPluginInstallation(
+	ctx context.Context, client client.Client, kpi *v1alpha1.KongPluginInstallation, msg string,
+) error {
+	return setStatusConditionForKongPluginInstallation(ctx, client, kpi, metav1.ConditionFalse, v1alpha1.KongPluginInstallationReasonFailed, msg)
+}
+
+func setStatusConditionForKongPluginInstallation(
+	ctx context.Context, client client.Client, kpi *v1alpha1.KongPluginInstallation, conditionStatus metav1.ConditionStatus, reason v1alpha1.KongPluginInstallationConditionReason, msg string,
+) error {
+	status := metav1.Condition{
+		Type:               string(v1alpha1.KongPluginInstallationConditionStatusAccepted),
+		Status:             conditionStatus,
+		ObservedGeneration: kpi.Generation,
+		LastTransitionTime: metav1.Now(),
+		Reason:             string(reason),
+		Message:            msg,
+	}
+	_, index, found := lo.FindIndexOf(kpi.Status.Conditions, func(c metav1.Condition) bool {
+		return c.Type == string(v1alpha1.KongPluginInstallationConditionStatusAccepted)
+	})
+	if found {
+		kpi.Status.Conditions[index] = status
+	} else {
+		kpi.Status.Conditions = append(kpi.Status.Conditions, status)
+	}
+	return client.Status().Update(ctx, kpi)
```

Comment:
I think this update will always cause a change in the resource status as we are updating the `lastTransitionTime`. We should check that there are some differences between the old and the new status, instead of updating it regardless.